framework/web/UrlRule.php

```diff
@@ -367,8 +367,15 @@ public function createUrl($manager, $route, $params)
                 continue;
             }
             if (!isset($params[$name])) {
-                return false;
-            } elseif (strcmp($params[$name], $value) === 0) { // strcmp will do string conversion automatically
+                // allow omit empty optional params
+                // @see https://github.com/yiisoft/yii2/issues/10970
+                if (in_array($name, $this->placeholders) && strcmp($value, '') === 0) {
```

Comment:
`$value` could be `0`. `empty()` will treat `0` the same as `''` (empty string) which is incorrect, and `===` will return fail when comparing `0` (integer) and `"0"` (string), which is impractical - finally all params are converted to string, so type is irrelevant. String comparison is the best option here.